### PR TITLE
Rclone Incremental backups

### PR DESCRIPTION
**Pull request by nickchomey**
I have overridden the SFTP IncBackup mechanism in order to add rclone to the incremental backup functionality, which will enable users to use over [40 different cloud storage providers](https://rclone.org/overview/). 

However, it will probably be necessary for you to create a new backup mechanism that adds:
* rclone item in the Select Type dropdown for Add/Remove Destinations, which presents text boxes for the rclone remote name and desired path in that remote
* ability to add multiple rclone remotes (OneDrive, Dropbox, Box, Google Drive etc...)
* dropdown in the Create Backup for which rclone remote to use - it seems reasonable to expect that someone would want to create backups of the same website to multiple cloud providers for redundancy.

Right now, the problem is with the `self.externalApp`, which makes rclone look for its `/.config/rclone/rclone.conf` file in the website's directory, which only exists if you specifically create the rclone remote while logged in with that website's user. This cannot work, as it doesn't make any sense to require the users to create a new remote for each cloud provider for each website - each remote should be created once for the entire server, either with the root user or cyberpanel user. 

I tried creating a symlink to `/.config/rclone/rclone.conf` within the website directory, but this didn't work. So, I suspect you will have to modify the mechanism to use some sort of environment variable that points rclone to `/.config/rclone/rclone.conf`

It is also possible that this raises other issues with regards to which remotes each user should have access to. I don't know enough about CyberPanel to know how the different users work, but I trust that you will know what the correct procedure will be.

### CODE REVIEW

File: IncBackups/IncBackupsControl.py [hunk 1]
```diff
@@ -43,6 +43,7 @@ def __init__(self, function, extraArgs):
         self.website = ''
         self.backupDestinations = ''
         self.jobid = 0
+        self.jobids = ''
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:46]: added to streamline the backup deletion process by sending all jobids at once (e.g. `restic -r remotename forget 2385238d 329349s9 23458s3j --prune --password-file... `)

File: IncBackups/IncBackupsControl.py [hunk 2]
```diff
@@ -184,8 +185,8 @@ def awsFunction(self, fType, backupPath=None, snapshotID=None, bType=None, delet
 
                     key, secret = self.getAWSData()
 
-                    command = 'export AWS_ACCESS_KEY_ID=%s AWS_SECRET_ACCESS_KEY=%s  && restic -r s3:s3.amazonaws.com/%s forget %s --password-file %s' % (
-                        key, secret, self.website, snapshotID, self.passwordFile)
+                    command = 'export AWS_ACCESS_KEY_ID=%s AWS_SECRET_ACCESS_KEY=%s  && restic -r s3:s3.amazonaws.com/%s forget %s --prune --password-file %s' % (
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:188]: combined the forget and prune mechanisms, and changed it so all jobids are forgotten at once (e.g. `restic -r remotename forget 2385238d 329349s9 23458s3j --prune --password-file... `) and a single prune is done afterwards - the saves considerable time because prune is VERY expensive

File: IncBackups/IncBackupsControl.py [hunk 3]
```diff
@@ -254,7 +250,7 @@ def localFunction(self, backupPath, type, restore=None, delete=None):
 
             repoLocation = '/home/%s/incbackup' % (self.website)
 
-            command = 'restic -r %s forget %s --password-file %s' % (repoLocation, self.jobid.snapshotid, self.passwordFile)
+            command = 'restic -r %s forget %s --prune --password-file %s' % (repoLocation, self.jobids, self.passwordFile)
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:253]: same thing - one forget/prune for all jobids

File: IncBackups/IncBackupsControl.py [hunk 4]
```diff
@@ -738,44 +728,42 @@ def metaBackup(self):
             return 0
 
     def initiateRepo(self):
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:730]: changed this whole mechanism.

First it does a check for the restic repo via checking for existing snapshots, and then only proceeds to create one if it doesn't exist. This was done per the instructions in their documentation: https://restic.readthedocs.io/en/latest/075_scripting.html

But I now realize that perhaps this isn't necessary and is more efficient to just try to init rather than check snapshots. The problem was that it was throwing errors because your error checking mechanism wasn't catching the message properly.

I also consolidated redundant code from within each of the local/sftp/aws if statements to after the if statements.

Finally I changed what gets logged - saying checking for repo, then either OK or repo initiated.

File: IncBackups/IncBackupsControl.py [hunk 5]
```diff
@@ -811,9 +799,9 @@ def createBackup(self):
 
             ### Checking if restic is installed before moving on
 
-            command = 'restic'
+            command = 'restic version'
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:802]: this generates a much smaller output in the log file

File: IncBackups/IncBackupsControl.py [hunk 6]
```diff
@@ -921,24 +909,21 @@ def DeleteSnapShot(self, inc_job):
...
             self.externalApp = job_snapshots[0].job.website.externalApp
             self.passwordFile = '/home/%s/%s' % (self.website, self.website)
-
             for job_snapshot in job_snapshots:
 
-                ## Functions above use the self.jobid varilable to extract information about this snapshot, so this below variable needs to be set
-
+                ## Functions above use the self.jobid variable to extract information about this snapshot, so this below variable needs to be set
                 self.jobid = job_snapshot
+                self.jobids = self.jobids + job_snapshot.snapshotid + ' '
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:916]: this is where I concatenate all of the jobids into one string, which is the passed only 1 time to the delete function. Considerable time savings

File: IncBackups/IncBackupsControl.py [hunk 7]
```diff
@@ -283,17 +276,17 @@ def localFunction(self, backupPath, type, restore=None, delete=None):
     ## Last argument delete is set when the snapshot is to be deleted from this repo, when this argument is set, any preceding argument is not used
 
     def sftpFunction(self, backupPath, type, restore=None, delete=None):
```
Comment (nickchomey) [inline IncBackups/IncBackupsControl.py:278]: This function was changed a lot to accomodate rclone (specifically onedrive right now, but that only refers to the name of the rclone remote that I created ("onedrive"). You could test this with any provider, so long as you call the rclone remote "onedrive".

**Comment by usmannasir** [comment]
Please commit this to `v2.1.2-dev`, because can't merge to stable as it will go live. Once you do that, I will be able to pull and test.

**Comment by nickchomey** [comment]
Further ideas for improvement of the Incremental Backup functionality

- [ ] Move the restic prune function to its own function, separate from forget. This can be run automatically once a day, rather than with every backup deletion. It will allow you to rapidly delete many backups, rather than having to wait a few minutes for each backup deletion
- [ ] Allow to set the frequency in minutes, hours and days for custom incremental backup schedule. I would like to do it every hour, because it doesn't create much additional data when there are no changes. 
- [ ] Add comments/labels to each backup, so that when we make a manual backup, we can remember why it was done
- [ ] Add ability to add multiple rclone remotes (OneDrive, Google Drive, Box, DropBox, etc...) in the Add/Remove Destinations tool
- [ ] Ability to use any/all of the remotes for each website, for redundancy
- [ ] Add branching to the backups/restores. If I restore a backup from 2 days ago, future backups will be part of that branch and it should be clear that the backups from yesterday and today are not part of that branch.
- [ ] Put the meta.xml, .sql and restic password files in a folder rather than in the root of /home/websitedirectory
- [ ] Documentation about Backup/Restore best practices. What happens if you restore data but not the database? Or database without data? I suspect both break the site, since plugin/file changes are, at the very least, reflected in wp_options. 
- [ ] CyberCP/plogical/randomPassword.py shoudl use a longer password. `openssl rand 60 | openssl base64 -A` is an option
- [ ] Add manual folder exclusions - for example, I user the Duplicator plugin which stores large backup files in wp-content/backups-dup-lite. These aren't necessary to backup again
- [ ] Ability to mount rclone remotes as a drive? Not sure how people would use this, but it could be helpful in some way. It is not difficult to do via SSH, not sure how it would work with CyberPanel

**Comment by nickchomey** [comment]
> Please commit this to `v2.1.2-dev`, because can't merge to stable as it will go live. Once you do that, I will be able to pull and test.

Done

**Comment by usmannasir** [comment]
It will take me sometime to review and check, pull merged for now.

File: IncBackups/IncBackupsControl.py [hunk 8]
```diff
@@ -57,9 +58,9 @@ def run(self):
             self.restorePoint()
 
     def getRemoteBackups(self):
-        if self.backupDestinations[:4] == 'sftp':
-            path = '/home/backup/%s' % (self.website)
-            command = 'export RESTIC_PASSWORD=%s PATH=${PATH}:/usr/bin && restic -r %s:%s snapshots' % (
+        if self.backupDestinations == 'onedrive':
```
Comment (usmannasir) [inline IncBackups/IncBackupsControl.py:61]: why did you change this to onedrive and not rclone?

Comment (nickchomey) [reply IncBackups/IncBackupsControl.py:61]: I explained this in a variety of comments here and in issue #834. 

This line refers not to rclone, but to the name of the rclone remote that you want to use. In this case, I was only setting it up to work with onedrive. Ideally we will change it so that it reads `self.backupDestinations = selectedRcloneRemote`. And that variable will contain the name of the rclone remote that has been added as a backup destination and was selected in the create backup window. But we don't have that mechanism yet, so I just hardcoded it as onedrive. 

For now, you can call it anything - onedrive, rclone, etc... The important part is that you can test this whole mechanism with ANY rclone provider, IF you name the remote "onedrive" (or whatever the string is in this line. 

For example, if you change it to "rclone", you would have to create a rclone remote with your desired cloud provider (google drive, onedrive, dropbox, etc) named rclone.

Does this make sense? It's just a placeholder to make it work until a mechanism is made for the front end to add multiple rclone remotes (that were created in ssh) as destinations for Inc backups, and to select them from the create backup page.

Also, once again, I've overridden the entire SFTP mechanism, which I assume you don't want to do. It's just for testing purposes and you can use it as an example to create a separate rclone mechanism alongside local, SFTP and aws. Or, if you think it is better, you can replace some or all of those with rclone. 

The only drawback of rclone is that setting up a remote probably requires some manual action for each provider. For example, for onedrive you have to run a command on your local computer to open the browser, log into your account and then copy and paste the authorization code back into the the rclone remote creation tool in ssh. I assume something similar needs to be done for most providers. If you can figure out how to write all of that to a textbox in the front end, that would be ideal. 

Though, every cyberpanel user should be familiar with basic ssh, so it really shouldn't be a problem to create the remotes via ssh and then just enter the remote name in a textbox in the Add Destination page. Especially if we document it clearly.
